universum/configuration_support.py

```diff
@@ -1,18 +1,210 @@
+        self.artifacts: str = artifacts
+        self.report_artifacts: str = report_artifacts
+        self.artifact_prebuild_clean: bool = artifact_prebuild_clean
+        self.critical: bool = critical
+        self.background: bool = background
+        self.finish_background: bool = finish_background
+        self.pass_tag: str = pass_tag
+        self.fail_tag: str = fail_tag
+        self.if_env_set: str = if_env_set
+        self.children: Optional['Variations'] = None
+        self.extras: Dict[str, str] = extras if extras else {}
+        for key, value in kwargs.items():
+            self.extras[key] = value
+
+    def __repr__(self) -> str:
+        """
+        This function simulates dict-like legacy representation for output
+        :return: dict-like string
+
+        >>> cfg = ProjectConfiguration(name='foo', command=['bar'], my_var='baz')
+        >>> repr(cfg)
+        "{'name': 'foo', 'command': 'bar', 'my_var': 'baz'}"
+        """
+        res = {k: v for k, v in self.__dict__.items() if v and k != 'extras'}
+        res.update(self.extras)
+        if len(self.command) == 1:  # command should be printed as one string, instead of list
+            res['command'] = self.command[0]
+        return str(res)
+
+    def __eq__(self, other: Any) -> bool:
+        """
+        This functions simulates dict-like legacy comparison
+
+        :param other: dict to compare values, or object to check exact equality
+        :return: 'True' if 'other' matches
+
+        >>> cfg1 = ProjectConfiguration(name='foo', my_var='bar')
+        >>> cfg2 = ProjectConfiguration(name='foo', my_var='bar')
+        >>> cfg3 = ProjectConfiguration(name='foo', my_var='bar', critical=True)
+        >>> cfg1 == cfg1
+        True
+        >>> cfg1 == cfg2
+        True
+        >>> cfg1 == cfg3
+        False
+        >>> cfg1 == {'name': 'foo', 'my_var': 'bar'}
+        True
+        >>> cfg1 == {'name': 'foo', 'my_var': 'bar', 'critical': False}
+        True
+        >>> cfg1 == {'name': 'foo', 'my_var': 'bar', 'test': None}
+        True
+        >>> cfg1 == {'name': 'foo', 'my_var': 'bar', 'test': ' '}
+        False
+        """
+        if isinstance(other, ProjectConfiguration):
+            return self == other.__dict__
+        if isinstance(other, dict):
+            for key, val in other.items():
+                if val and self[key] != val:
+                    return False
+            return True
+        return super().__eq__(other)
+
+    def __getitem__(self, item: str) -> Optional[str]:
+        """
+        This functions simulates dict-like legacy access
+
+        :param item: client-defined item
+        :return: client-defined value
+
+        >>> cfg = ProjectConfiguration(name='foo', my_var='bar')
+        >>> cfg['name']
+        'foo'
+        >>> cfg['my_var']
+        'bar'
+        >>> cfg['test']
+        """
+        return self.extras.get(item, self.__dict__.get(item, None))
```

Comment:
From this implementation it looks like `self.extras` is kind of override for `self.__dict__`.
As can be seen in `self.__init__` there are two ways to add items into "self.extras": from explicitly passed "extras" argument and from keyword dictionary (kwargs). The first way allows to actually override "native" keys.
I my opinion there are several issues here:
 - do we need such overriding at all?
 - are we using "extras" argument anywhere?
 - "extras" work only for dict-like access (cfg['my_var']), but they don't work for object-like access (cfg.my_var)

Reply:
"extras" exists to support client-defined dict values. With extras as it is - clients should notice no changes when this patch is applied. Internally universum should never use extras.
You're right that case when "extras" is a dict key deserves special attention - I'll add more tests for that case.

Reply:
I don't think "extras" need to support object-like access, since they only exist to simulate legacy behaviour. I think best way would be to generate a warning anytime extras is used.